controllers/pagerdutyintegration/serviceorchestration_handler.go

```diff
@@ -116,29 +115,33 @@ func (r *PagerDutyIntegrationReconciler) handleServiceOrchestration(pdclient pd.
 		return err
 	}
 
-	if pdiSelector.Matches(labels.Set(ruleConfigmap.GetLabels())) {
-		orchestrationConfig, err := utils.LoadConfigMapData(r.Client,
+	var orchestrationRuleConfigData string
+	if pdiSelector.Matches(labels.Set(orchestrationRuleConfigmap.GetLabels())) {
+		orchestrationRuleConfigData, err = utils.LoadConfigMapData(r.Client,
 			serviceOrchestrationConfigMap, ServiceOrchestrationDataName)
 		if err != nil {
 			return err
 		}
 
-		pdData.ServiceOrchestrationRules = orchestrationConfig
-
-		r.reqLogger.Info(fmt.Sprintf("apply the service orchestration rules from configmap: %s",
-			orchestrationConfigmapName))
-		err = pdclient.ApplyServiceOrchestrationRule(pdData)
-		if err != nil {
-			return err
-		}
-
-		pdData.ServiceOrchestrationRuleApplied = true
-
-		err = pdData.SetClusterConfig(r.Client, cd.Namespace, clusterConfigmapName)
-		if err != nil {
-			r.reqLogger.Error(err, "Error updating PagerDuty cluster config", "Name",
-				clusterConfigmapName)
-			return err
+		if pdData.ServiceOrchestrationRuleApplied != orchestrationRuleConfigData {
+			// Apply rule for new service
+			pdData.ServiceOrchestrationRuleApplied = orchestrationRuleConfigData
+
+			r.reqLogger.Info(fmt.Sprintf("applying the service orchestration rules from configmap: %s",
+				orchestrationConfigmapName))
+			err = pdclient.ApplyServiceOrchestrationRule(pdData)
```

Comment:
Just wondering, as I've not tested the code - does this result in the existing rule(s) being entirely replaced with the new rule? Just want to make sure this doesn't result in a second new rule being added to the service's ruleset whilst leaving the first still there.

I ask because the PD API [0] is unclear to me when I read it as to whether the orchestration rule id is required in order to update-in-place an existing rule. And the id would not be known, nor passed through to the PUT, because it's using the definition in the orchestration rule configmap for the body of the PUT.

[0] https://developer.pagerduty.com/api-reference/aeaec09888036-update-the-service-orchestration-for-a-service

Reply:
Yes, it will update the existing rule (not sure if we should call it existing rule, but there will be only one service orchestration rule for each service.)

Reply:
Excellent, thanks for checking!